tests.system.reader.test_reader: test_snapshot[v1] failed

This test failed!

To configure my behavior, see [the Flaky Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/master/packages/flakybot).

If I'm commenting on this issue too often, add the `flakybot: quiet` label and
I will stop commenting.

---

commit: f1bc4128e099e83d5a15408bd6fd2129c2151301
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/679dd086-80e7-40bf-a5e8-dfd64c0c6d62), [Sponge](http://sponge2/679dd086-80e7-40bf-a5e8-dfd64c0c6d62)
status: failed
<details><summary>Test output</summary><br><pre>client_and_types = (<google.cloud.bigquery_storage_v1.BigQueryReadClient object at 0x7f8b42171fd0>, <module 'google.cloud.bigquery_storage_v1.gapic_types' from '/tmpfs/src/github/python-bigquery-storage/google/cloud/bigquery_storage_v1/gapic_types.py'>)
project_id = 'precise-truck-742'
table_with_data_ref = 'projects/precise-truck-742/datasets/bq_storage_system_tests_42335d05_507f_48e0_8c32_033823270111/tables/users'
bq_client = <google.cloud.bigquery.client.Client object at 0x7f8b42149ca0>

    def test_snapshot(client_and_types, project_id, table_with_data_ref, bq_client):
        client, types = client_and_types
        before_new_data = dt.datetime.now(tz=dt.timezone.utc)
    
        # load additional data into the table
        new_data = [
            {"first_name": "NewGuyFoo", "last_name": "Smith", "age": 46},
            {"first_name": "NewGuyBar", "last_name": "Jones", "age": 30},
        ]
    
        destination = _to_bq_table_ref(table_with_data_ref)
        bq_client.load_table_from_json(new_data, destination).result()
    
        # read data using the timestamp before the additional data load
    
        read_session = types.ReadSession()
        read_session.table = table_with_data_ref
        read_session.table_modifiers.snapshot_time = before_new_data
        read_session.data_format = types.DataFormat.AVRO
    
        session = client.create_read_session(
            request={
                "parent": "projects/{}".format(project_id),
                "read_session": read_session,
                "max_stream_count": 1,
            }
        )
        stream = session.streams[0].name
    
        rows = list(client.read_rows(stream).rows(session))
    
        # verify that only the data before the timestamp was returned
>       assert len(rows) == 5  # all initial records
E       assert 0 == 5
E        +  where 0 = len([])

tests/system/reader/test_reader.py:212: AssertionError</pre></details>

```
nox > py.test --quiet --junitxml=system_3.8_sponge_log.xml tests/system
.........F................................                               [100%]
=================================== FAILURES ===================================
______________________________ test_snapshot[v1] _______________________________

client_and_types = (<google.cloud.bigquery_storage_v1.BigQueryReadClient object at 0x7f8b42171fd0>, <module 'google.cloud.bigquery_storage_v1.gapic_types' from '/tmpfs/src/github/python-bigquery-storage/google/cloud/bigquery_storage_v1/gapic_types.py'>)
project_id = 'precise-truck-742'
table_with_data_ref = 'projects/precise-truck-742/datasets/bq_storage_system_tests_42335d05_507f_48e0_8c32_033823270111/tables/users'
bq_client = <google.cloud.bigquery.client.Client object at 0x7f8b42149ca0>

    def test_snapshot(client_and_types, project_id, table_with_data_ref, bq_client):
        client, types = client_and_types
        before_new_data = dt.datetime.now(tz=dt.timezone.utc)
    
        # load additional data into the table
        new_data = [
            {"first_name": "NewGuyFoo", "last_name": "Smith", "age": 46},
            {"first_name": "NewGuyBar", "last_name": "Jones", "age": 30},
        ]
    
        destination = _to_bq_table_ref(table_with_data_ref)
        bq_client.load_table_from_json(new_data, destination).result()
    
        # read data using the timestamp before the additional data load
    
        read_session = types.ReadSession()
        read_session.table = table_with_data_ref
        read_session.table_modifiers.snapshot_time = before_new_data
        read_session.data_format = types.DataFormat.AVRO
    
        session = client.create_read_session(
            request={
                "parent": "projects/{}".format(project_id),
                "read_session": read_session,
                "max_stream_count": 1,
            }
        )
        stream = session.streams[0].name
    
        rows = list(client.read_rows(stream).rows(session))
    
        # verify that only the data before the timestamp was returned
>       assert len(rows) == 5  # all initial records
E       assert 0 == 5
E        +  where 0 = len([])

tests/system/reader/test_reader.py:212: AssertionError
```

Query returned 0 rows. Perhaps we need to adjust the timestamps we use for the snapshot?

--- a/tests/system/conftest.py
+++ b/tests/system/conftest.py
@@ -88,7 +88,7 @@ def dataset(project_id, bq_client):
     bq_client.delete_dataset(dataset, delete_contents=True)
 
 
-@pytest.fixture(scope="session")
+@pytest.fixture
 def table(project_id, dataset, bq_client):
     from google.cloud import bigquery
 

--- a/tests/system/reader/test_reader.py
+++ b/tests/system/reader/test_reader.py
@@ -19,12 +19,14 @@ import copy
 import datetime as dt
 import decimal
 import re
-import time
 
 import pytest
 import pytz
 
 from google.cloud import bigquery
+
+
+_TABLE_FORMAT = "projects/{}/datasets/{}/tables/{}"
 
 
 def _to_bq_table_ref(table_name_string, partition_suffix=""):
@@ -178,30 +180,37 @@ def test_column_selection_read(
         assert sorted(row.keys()) == ["age", "first_name"]
 
 
-def test_snapshot(client_and_types, project_id, table_with_data_ref, bq_client):
+@pytest.mark.parametrize("data_format", ("AVRO", "ARROW"))
+def test_snapshot(client_and_types, project_id, table, bq_client, data_format):
     client, types = client_and_types
-    # Sleep for a moment to give use some wiggle room in case the BigQuery
-    # snapshot time and our times are out-of-sync.
-    # https://github.com/googleapis/python-bigquery-storage/issues/151
-    time.sleep(10)
-    before_new_data = dt.datetime.now(tz=dt.timezone.utc)
-    time.sleep(10)
+
+    # load original data into the table
+    original_data = [
+        {"first_name": "OGFoo", "last_name": "Smith", "age": 44},
+        {"first_name": "OGBar", "last_name": "Jones", "age": 33},
+    ]
+    og_job = bq_client.load_table_from_json(original_data, table).result()
+    og_time = og_job.ended
 
     # load additional data into the table
     new_data = [
-        {"first_name": "NewGuyFoo", "last_name": "Smith", "age": 46},
-        {"first_name": "NewGuyBar", "last_name": "Jones", "age": 30},
+        {"first_name": "NewFoo", "last_name": "Smiff", "age": 43},
+        {"first_name": "NewBar", "last_name": "Jomes", "age": 34},
     ]
+    new_job = bq_client.load_table_from_json(new_data, table).result()
+    new_time = new_job.ended
 
-    destination = _to_bq_table_ref(table_with_data_ref)
-    bq_client.load_table_from_json(new_data, destination).result()
+    # Because we want our snapshot to be between when we loaded the original
+    # data and when the new data was loaded, take the average of the two load
+    # job completion times.
+    before_new_data = og_time + ((new_time - og_time) / 2)
 
     # read data using the timestamp before the additional data load
-
+    table_path = _TABLE_FORMAT.format(table.project, table.dataset_id, table.table_id)
     read_session = types.ReadSession()
-    read_session.table = table_with_data_ref
+    read_session.table = table_path
     read_session.table_modifiers.snapshot_time = before_new_data
-    read_session.data_format = types.DataFormat.AVRO
+    read_session.data_format = data_format
 
     session = client.create_read_session(
         request={
@@ -215,10 +224,11 @@ def test_snapshot(client_and_types, project_id, table_with_data_ref, bq_client):
     rows = list(client.read_rows(stream).rows(session))
 
     # verify that only the data before the timestamp was returned
-    assert len(rows) == 5  # all initial records
+    assert len(rows) == 2  # all initial records
 
     for row in rows:
-        assert "NewGuy" not in row["first_name"]  # no new records
+        assert "OG" in str(row["first_name"])
+        assert "New" not in str(row["first_name"])
 
 
 def test_column_partitioned_table(

--- a/tests/system/reader/test_reader.py
+++ b/tests/system/reader/test_reader.py
@@ -19,6 +19,7 @@ import copy
 import datetime as dt
 import decimal
 import re
+import time
 
 import pytest
 import pytz
@@ -179,6 +180,10 @@ def test_column_selection_read(
 
 def test_snapshot(client_and_types, project_id, table_with_data_ref, bq_client):
     client, types = client_and_types
+    # Sleep for a moment to give use some wiggle room in case the BigQuery
+    # snapshot time and our times are out-of-sync.
+    # https://github.com/googleapis/python-bigquery-storage/issues/151
+    time.sleep(30)
     before_new_data = dt.datetime.now(tz=dt.timezone.utc)
 
     # load additional data into the table
@@ -189,6 +194,7 @@ def test_snapshot(client_and_types, project_id, table_with_data_ref, bq_client):
 
     destination = _to_bq_table_ref(table_with_data_ref)
     bq_client.load_table_from_json(new_data, destination).result()
+    time.sleep(30)
 
     # read data using the timestamp before the additional data load
 

--- a/tests/system/reader/test_reader.py
+++ b/tests/system/reader/test_reader.py
@@ -183,8 +183,9 @@ def test_snapshot(client_and_types, project_id, table_with_data_ref, bq_client):
     # Sleep for a moment to give use some wiggle room in case the BigQuery
     # snapshot time and our times are out-of-sync.
     # https://github.com/googleapis/python-bigquery-storage/issues/151
-    time.sleep(30)
+    time.sleep(10)
     before_new_data = dt.datetime.now(tz=dt.timezone.utc)
+    time.sleep(10)
 
     # load additional data into the table
     new_data = [
@@ -194,7 +195,6 @@ def test_snapshot(client_and_types, project_id, table_with_data_ref, bq_client):
 
     destination = _to_bq_table_ref(table_with_data_ref)
     bq_client.load_table_from_json(new_data, destination).result()
-    time.sleep(30)
 
     # read data using the timestamp before the additional data load
 

